cms/djangoapps/contentstore/views/assets.py

```diff
@@ -94,97 +109,168 @@ def _assets_json(request, course_key):
+        assets, total_count = _get_assets_for_page(request, course_key, query_options)
+
+
+    last_asset_to_display_index = first_asset_to_display_index + len(assets)
+    assets_in_json_format = _get_assets_in_json_format(assets, course_key)
+
+    response_payload = {
+        'start': first_asset_to_display_index,
+        'end': last_asset_to_display_index,
         'page': current_page,
         'pageSize': requested_page_size,
         'totalCount': total_count,
-        'assets': asset_json,
-        'sort': requested_sort,
-    })
+        'assets': assets_in_json_format,
+        'sort': request_options['requested_sort'],
+    }
+
+    return JsonResponse(response_payload)
+
+
+def _parse_request_to_dictionary(request):
+
+    return {
+        'requested_page': int(_get_requested_attribute(request, 'page')),
+        'requested_page_size': int(_get_requested_attribute(request, 'page_size')),
+        'requested_sort': _get_requested_attribute(request, 'sort'),
+        'requested_sort_direction': _get_requested_attribute(request, 'direction'),
+        'requested_asset_type': _get_requested_attribute(request, 'asset_type')
+    }
+
+
+def _get_requested_attribute(request, attribute):
+
+    return request.GET.get(attribute, request_defaults.get(attribute))
+
+
+def _get_filter_parameters_for_mongo(requested_filter):
+
+    if requested_filter == 'OTHER':
+        mongo_where_operator_parameters = _get_mongo_where_operator_parameters_for_other()
+    else:
+        mongo_where_operator_parameters = _get_mongo_where_operator_parameters_for_filters(requested_filter)
+    return mongo_where_operator_parameters
+
+
+def _get_mongo_where_operator_parameters_for_other():
+
+    requested_file_types = _get_files_and_upload_type_filters().keys()
+    file_extensions_for_requested_file_types = _get_extensions_for_file_types(requested_file_types)
+    javascript_expression_to_filter_extensions = _get_javascript_expressions_to_filter_extensions_with_operator(file_extensions_for_requested_file_types, "!=")
+    javascript_expressions_to_filter_extensions_in_mongo = _get_javascript_expressions_for_mongo_filter_with_separator(javascript_expression_to_filter_extensions, ' && ')
+
+    return javascript_expressions_to_filter_extensions_in_mongo
+
+
+def _get_mongo_where_operator_parameters_for_filters(requested_filter):
+
+    requested_file_types = _get_requested_file_types_from_request(requested_filter)
+    file_extensions_for_request_file_types = _get_extensions_for_file_types(requested_file_types)
+    javascript_expressions_to_filter_extensions = _get_javascript_expressions_to_filter_extensions_with_operator(file_extensions_for_requested_file_types, "==")
+    javascript_expressions_to_filter_extensions_in_mongo = _get_javascript_expressions_for_mongo_filter_with_separator(javascript_expressions_to_filter_extensions, ' || ')
+
+    return javascript_expressions_to_filter_extensions_in_mongo
```

Comment:
So, I have an idea about this whole section (`_get_mongo_where_operator_parameters_for_other` and `_get_mongo_where_operator_parameters_for_filters`; the `&&` vs `||` and `==` vs `!=` discussions), feel free to ignore if you don't like it.

The query strings we're building are either `extension != '.abc' && extension != '.def' && ...` or `extension == '.abc' || extension == '.def' || ...`. I'd like to suggest that instead of coding the separators and operators, we make clever use of [De Morgan's Laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws). We can write both query strings as `extension != '.abc' && extension != '.def' && ...` (or the `==` and `||` version, as long as we standardize to one or the other), with one being the logical negation of the other by adding a `NOT` outside the whole query.

Then, `_get_filter_parameters_for_mongo` can be much simpler, as the "inner query string" is the same in both cases, just with a specified list of filters vs all possible filters.